python/hail/expr/functions.py

```diff
@@ -1779,6 +1779,110 @@ def sqrt(x) -> Float64Expression:
     return _func("sqrt", tfloat64, x)
 
 
+@typecheck(s1=expr_str, s2=expr_str)
+def hamming(s1, s2) -> Int32Expression:
+    """Returns the Hamming distance between the two strings.
+
+    Examples
+    --------
+    .. doctest::
+
+        >>> hl.eval_expr(hl.hamming('ATATA', 'ATGCA'))
+        2
+
+        >>> hl.eval_expr(hl.hamming('abcdefg', 'zzcdefz'))
+        3
+
+    Notes
+    -----
+    This method will fail if the two strings have different length.
+
+    Parameters
+    ----------
+    s1 : :class:`.StringExpression`
+        First string.
+    s2 : :class:`.StringExpression`
+        Second string.
+
+    Returns
+    -------
+    :class:`.Expression` of type :py:data:`.tint32`
+    """
+    return _func("hamming", tint32, s1, s2)
+
+
+@typecheck(s=expr_str)
+def is_valid_allele(s) -> BooleanExpression:
+    return s.matches(r'^([ACGT]+)|\\*$')
+
+
+_allele_enum = {i: v for i, v in enumerate(["invalid", "SNP", "MNP", "Insertion", "Deletion", "Complex", "Star"])}
+_allele_ints = {v: k for k, v in _allele_enum.items()}
+
+
+@typecheck(ref=expr_str, alt=expr_str)
+def _num_allele_type(ref, alt) -> Int32Expression:
+    return hl.bind(lambda r, a:
+                   hl.cond(is_valid_allele(r) & is_valid_allele(a),
+                           hl.case()
+                           .when(r.length() == a.length(),
+                                        hl.cond(r.length() == 1,
+                                                hl.cond((r == "*") | (a == "*"),
+                                                        _allele_ints["Star"],
+                                                        _allele_ints["SNP"]),
+                                                hl.cond(hamming(r, a) == 1,
+                                                        _allele_ints["SNP"],
+                                                        _allele_ints["MNP"])
+                                                )
+                                        )
+                           .when((r == "*") | (a == "*"), _allele_ints["Star"])
+                           .when((r.length() < a.length()) & (r[0] == a[0]) & a.endswith(r[1:]),
+                                 _allele_ints["Insertion"])
+                           .when((r[0] == a[0]) & r.endswith(a[1:]),
+                                 _allele_ints["Deletion"])
+                           .default(_allele_ints["Complex"]),
+                           _allele_ints["invalid"]),
```

Comment:
maybe prefer missing to invalid? I dunno, though

Reply:
the old behavior throws an error. I haven't dealt with that here because I'm not sure how to throw an error here :-| but it's not missing, it's an invalid ref/alt pair.

Reply:
I'm in favor of not throwing an error, but I think a missing allele type is nicer than than "invalid". I'm not sure though.

Reply:
I think we should change `"invalid"` to `"Unknown"`. That makes it clear that it is not a missing value -- we just don't have an appropriate label for the type of variant it is. I agree we shouldn't throw a fatal error.

Reply:
nice. I'll do that.